src/Libraries/Microsoft.Extensions.AI/ChatCompletion/ChatClientBuilder.cs

```diff
@@ -76,4 +78,62 @@ public ChatClientBuilder Use(Func<IChatClient, IServiceProvider, IChatClient> cl
         (_clientFactories ??= []).Add(clientFactory);
         return this;
     }
+
+    /// <summary>
+    /// Adds to the chat client pipeline an anonymous delegating chat client based on a delegate that provides
+    /// an implementation for both <see cref="IChatClient.CompleteAsync"/> and <see cref="IChatClient.CompleteStreamingAsync"/>.
+    /// </summary>
+    /// <param name="sharedFunc">
+    /// A delegate that provides the implementation for both <see cref="IChatClient.CompleteAsync"/> and
+    /// <see cref="IChatClient.CompleteStreamingAsync"/>. In addition to the arguments for the operation, it's
+    /// provided with a delegate to the inner client that should be used to perform the operation on the inner client.
+    /// It will handle both the non-streaming and streaming cases.
+    /// </param>
+    /// <returns>The updated <see cref="ChatClientBuilder"/> instance.</returns>
+    /// <remarks>
+    /// This overload may be used when the anonymous implementation needs to provide pre- and/or post-processing, but doesn't
+    /// need to interact with the results of the operation, which will come from the inner client.
+    /// </remarks>
+    /// <exception cref="ArgumentNullException"><paramref name="sharedFunc"/> is <see langword="null"/>.</exception>
+    public ChatClientBuilder Use(AnonymousDelegatingChatClient.CompleteSharedFunc sharedFunc)
```

Comment:
Sidenote: while this will be fine to use when copy-pasting an example, it will be challenging to use just through intellisense alone. AFAIK VS doesn't provide any prompt to indicate what set of args a lambda should take in order to conform to a delegate definition. Obviously there's nothing you can do about that here and it's fine for this PR.

But I wonder if we should consider asking for a C#/Intellisense improvement whereby, when the cursor is in a place where you're being asked for a delegate, it should include some sort of "generate lambda" option in the list of available completions which would inject code like `(T0 arg0, T1 arg1, T2 arg2) => { }`.

I remember in the past having this issue with Blazor's grid data provider callback. The delegate type is just too complicated to remember so I can never work out how to actually write the code for a grid data provider and have to go looking for an example to copy-paste.